### PR TITLE
shared memory histogram cache

### PR DESCRIPTION
**Pull request by psychocoderHPC**
Add Histogram shared memory cache

### CODE REVIEW

File: include/picongpu/particles/atomicPhysics2/RollElectronBin.hpp [hunk 1]
```diff
@@ -63,36 +63,32 @@ namespace picongpu::particles::atomicPhysics2::rollElectronBin
      * @param electronHistogram electron histogram
      * @param atomicData atomicData dataBoxes
      */
-    template<typename T_Histogram, typename T_RateFunctional, typename... T_AtomicData>
+    template<typename T_RateFunctional, typename T_Histogram, typename... T_AtomicData>
     HDINLINE static uint32_t findBin(
```
Comment (psychocoderHPC) [inline include/picongpu/particles/atomicPhysics2/RollElectronBin.hpp]: IMO the full class is not used anymore and can be deleted!

File: include/picongpu/particles/atomicPhysics2/rateCalculation/BoundBoundTransitionRates.hpp [hunk 2]
```diff
@@ -253,48 +254,40 @@ namespace picongpu::particles::atomicPhysics2::rateCalculation
             // [1e6b * (eV)^2]
 
             // 1e6b*(eV)^2 / (eV)^2 * unitless * (eV)/(eV) * unitless = 1e6b
-            float_X crossSection_butGaunt = constantPart / math::sqrt(3.)
+            float_X crossSection_butGaunt = constantPart / math::sqrt(3._X)
```
Comment (psychocoderHPC) [inline include/picongpu/particles/atomicPhysics2/rateCalculation/BoundBoundTransitionRates.hpp]: if `math::sqrt(3.)` is used we should cast the result to `float_X`

File: include/picongpu/simulation/stage/AtomicPhysics2.hpp [hunk 3]
```diff
@@ -221,13 +223,19 @@ namespace picongpu::simulation::stage
...
 
             setTimeRemaining(); // = (Delta t)_PIC
             ForEachIonSpeciesFixAtomicState{}(mappingDesc);
 
             uint16_t counterSubStep = 0u;
             // atomicPhysics sub-stepping loop, ends when timeRemaining<=0._X
+
+#if 1
+            if(currentStep >= 1u)
+                cudaProfilerStart();
```
Comment (psychocoderHPC) [inline include/picongpu/simulation/stage/AtomicPhysics2.hpp]: will be removed, only  useful for profiling

**Comment by psychocoderHPC** [comment]
I moved most refactorings to separate PRs, after all changes were merged only the cache histogram should be left.

**Comment by psychocoderHPC** [comment]
@BrianMarre ready for review

File: include/picongpu/particles/atomicPhysics2/electronDistribution/CachedHistogram.hpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+         * @attention This method is synchronizing the worker before returning the handle.
+         *
+         * @tparam T_Worker
+         * @tparam T_Histogram
+         * @param worker
+         * @param electronHistogram
+         * @param volumeScalingFactor
+         */
+        template<typename T_Worker, typename T_Histogram>
+        HDINLINE void fill(T_Worker const& worker, T_Histogram const& electronHistogram, float_X volumeScalingFactor)
```
Comment (BrianMarre) [inline include/picongpu/particles/atomicPhysics2/electronDistribution/CachedHistogram.hpp]: volumeScalingFactor should be const

File: include/picongpu/particles/atomicPhysics2/kernel/ChooseTransition_BoundBound.kernel [hunk 5]
```diff
@@ -225,22 +231,15 @@ namespace picongpu::particles::atomicPhysics2::kernel
...
                                 float_X const rateTransition = picongpu::particles::atomicPhysics2::rateCalculation::
                                     BoundBoundTransitionRates<T_n_max>::template rateCollisionalBoundBoundTransition<
                                         T_AtomicStateDataDataBox,
                                         T_BoundBoundTransitionDataBox,
                                         isUpward>(
-                                        energy,
-                                        binWidth,
+                                        cachedHistogram.energy[binIndex],
+                                        cachedHistogram.binWidth[binIndex],
                                         density,
```
Comment (BrianMarre) [inline include/picongpu/particles/atomicPhysics2/kernel/ChooseTransition_BoundBound.kernel:243]: why not an in place call?

**Review by BrianMarre — APPROVED** [review]

File: include/picongpu/particles/atomicPhysics2/kernel/ChooseTransition_BoundBound.kernel (hunk 5 above)
Comment (psychocoderHPC) [reply include/picongpu/particles/atomicPhysics2/kernel/ChooseTransition_BoundBound.kernel:243]: density is used in the debug message too

**Pull request merged**